File: tests/bench/08_combined.go
// Benchmark 08: Combined — realistic data processing pipeline
// Exercises: loops, arrays, maps, closures, filter, reduce, strings
package main

import (
	"fmt"
	"sort"
	"strings"
	"time"
)

func lcg(seed int64) int64 {
	return (seed*1103515245 + 12345) % 2147483648
}

type Record struct {
	Name     string
	Score    int64
	Category string
}

type Stats struct {
	Count int
	Avg   int64
	Min   int64
	Max   int64
}

func main() {
	nRecords := 500_000

	// Build dataset
	buildStart := time.Now()
	records := make([]Record, 0, nRecords)
	seed := int64(42)
	for i := 0; i < nRecords; i++ {
		seed = lcg(seed)
		score := seed % 100
		cat := score % 4
		var category string
		switch cat {
		case 0:
			category = "alpha"
		case 1:
			category = "beta"
		case 2:
			category = "gamma"
		default:
			category = "delta"
		}
		records = append(records, Record{
			Name:     fmt.Sprintf("item%d", i),
			Score:    score,
			Category: category,
		})
	}
	buildMs := time.Since(buildStart).Milliseconds()

	// Filter: keep records with score >= 50
	filtStart := time.Now()
	highScores := make([]Record, 0)
	for _, r := range records {
		if r.Score >= 50 {
			highScores = append(highScores, r)
		}
	}
	filtMs := time.Since(filtStart).Milliseconds()

	// Transform: double scores
	transStart := time.Now()
	transformed := make([]Record, len(highScores))
	for i, r := range highScores {
		transformed[i] = Record{
			Name:     r.Name,
			Score:    r.Score * 2,
			Category: r.Category,
		}
	}
	transMs := time.Since(transStart).Milliseconds()

	// Group by category
	groupStart := time.Now()
	groups := make(map[string][]int64)
	for _, r := range transformed {
		groups[r.Category] = append(groups[r.Category], r.Score)
	}
	groupMs := time.Since(groupStart).Milliseconds()

	// Sort each group and compute stats
	sortStart := time.Now()
	stats := make(map[string]Stats)
	for k, scores := range groups {
		sort.Slice(scores, func(i, j int) bool { return scores[i] < scores[j] })
		count := len(scores)
		sum := int64(0)
		for _, s := range scores {
			sum += s
		}
		avg := sum / int64(count)
		stats[k] = Stats{
			Count: count,
			Avg:   avg,
			Min:   scores[0],
			Max:   scores[count-1],
		}
	}
	sortMs := time.Since(sortStart).Milliseconds()

	// Format output
	fmtStart := time.Now()
	var sb strings.Builder
	statKeys := make([]string, 0, len(stats))
	for k := range stats {
		statKeys = append(statKeys, k)
	}
	sort.Strings(statKeys)
	for _, k := range statKeys {
		s := stats[k]
		sb.WriteString(fmt.Sprintf("%s: count=%d avg=%d min=%d max=%d\n", k, s.Count, s.Avg, s.Min, s.Max))
	}
	output := sb.String()
	fmtMs := time.Since(fmtStart).Milliseconds()

	totalMs := buildMs + filtMs + transMs + groupMs + sortMs + fmtMs
	fmt.Printf("[bench] combined: %d records processed in %d ms\n", nRecords, totalMs)
	fmt.Printf("[bench] combined_detail: build=%dms filter=%dms transform=%dms group=%dms sort=%dms format=%dms\n",
		buildMs, filtMs, transMs, groupMs, sortMs, fmtMs)
	fmt.Printf("[bench] combined_counts: total=%d filtered=%d groups=%d\n",
		len(records), len(transformed), len(statKeys))
	fmt.Print(output)
}

File: tests/bench/08_combined.rs
// Benchmark 08: Combined — realistic data processing pipeline
// Exercises: loops, arrays, maps, closures, iterators, filter, reduce, strings

use std::collections::HashMap;
use std::time::Instant;

fn lcg(seed: i64) -> i64 {
    (seed.wrapping_mul(1103515245).wrapping_add(12345)) % 2147483648
}

struct Record {
    name: String,
    score: i64,
    category: String,
}

struct Stats {
    count: usize,
    avg: i64,
    min: i64,
    max: i64,
}

fn main() {
    let n_records = 500_000;

    // Build dataset
    let build_start = Instant::now();
    let mut records: Vec<Record> = Vec::new();
    let mut seed: i64 = 42;
    for i in 0..n_records {
        seed = lcg(seed);
        let score = seed % 100;
        let cat = score % 4;
        let category = match cat {
            0 => "alpha",
            1 => "beta",
            2 => "gamma",
            _ => "delta",
        };
        records.push(Record {
            name: format!("item{}", i),
            score,
            category: category.to_string(),
        });
    }
    let build_ms = build_start.elapsed().as_millis();

    // Filter: keep records with score >= 50
    let filt_start = Instant::now();
    let high_scores: Vec<&Record> = records.iter().filter(|r| r.score >= 50).collect();
    let filt_ms = filt_start.elapsed().as_millis();

    // Transform: double scores
    let trans_start = Instant::now();
    let transformed: Vec<Record> = high_scores
        .iter()
        .map(|r| Record {
            name: r.name.clone(),
            score: r.score * 2,
            category: r.category.clone(),
        })
        .collect();
    let trans_ms = trans_start.elapsed().as_millis();

    // Group by category
    let group_start = Instant::now();
    let mut groups: HashMap<String, Vec<i64>> = HashMap::new();
    for r in &transformed {
        groups.entry(r.category.clone()).or_default().push(r.score);
    }
    let group_ms = group_start.elapsed().as_millis();

    // Sort each group and compute stats
    let sort_start = Instant::now();
    let mut stats: HashMap<String, Stats> = HashMap::new();
    for (k, scores) in &mut groups {
        scores.sort();
        let count = scores.len();
        let avg = scores.iter().sum::<i64>() / count as i64;
        let min = scores[0];
        let max = scores[count - 1];
        stats.insert(k.clone(), Stats { count, avg, min, max });
    }
    let sort_ms = sort_start.elapsed().as_millis();

    // Format output
    let fmt_start = Instant::now();
    let mut output = String::new();
    let mut stat_keys: Vec<&String> = stats.keys().collect();
    stat_keys.sort();
    for k in &stat_keys {
        let s = &stats[*k];
        output.push_str(&format!(
            "{}: count={} avg={} min={} max={}\n",
            k, s.count, s.avg, s.min, s.max
        ));
    }
    let fmt_ms = fmt_start.elapsed().as_millis();

    let total_ms = build_ms + filt_ms + trans_ms + group_ms + sort_ms + fmt_ms;
    println!("[bench] combined: {} records processed in {} ms", n_records, total_ms);
    println!(
        "[bench] combined_detail: build={}ms filter={}ms transform={}ms group={}ms sort={}ms format={}ms",
        build_ms, filt_ms, trans_ms, group_ms, sort_ms, fmt_ms
    );
    println!(
        "[bench] combined_counts: total={} filtered={} groups={}",
        records.len(),
        transformed.len(),
        stat_keys.len()
    );
    print!("{}", output);
}
